dask/dataframe/io/parquet/utils.py

```diff
@@ -0,0 +1,296 @@
+
+        This function is called many times in individual tasks
+
+        Parameters
+        ----------
+        fs: FileSystem
+        piece: object
+            This is some token that is returned by Engine.read_metadata.
+            Typically it represents a row group in a Parquet dataset
+        columns: List[str]
+            List of column names to pull out of that row group
+        partitions:
+        categories:
+
+        Returns
+        -------
+        A Pandas DataFrame
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def create_metadata(df, fs, path, append=False, **kwargs):
+        """Create engine-specific metadata instance for this dataset
+
+        Parameters
+        ----------
+        df: dask.dataframe.DataFrame
+        fs: FileSystem
+        path: str
+            Output file to write to, usually ``"_metadata"`` in the root of
+            the output dataset; only required if appending.
+        append: bool
+            If True, may use existing metadata (if any) and perform checks
+            against the new data being stored.
+
+        Returns
+        -------
+        tuple:
+            engine-specific instance
+            list of filenames, one per partition
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def write_metadata(parts, meta, fs, path, append=False, **kwargs):
+        """
+        Write a Dask DataFrame to Parquet
+
+        Parameters
+        ----------
+        parts: List
+            Containes metadata objects to write, of the type undrestood by the
+            specific implementation
+        meta: non-chunk metadata
+            Details that do not depend on the specifics of each chunk write,
+            typically the schema and pandas metadata, in a format the writer
+            can use.
+        fs: FileSystem
+        path: str
+            Output file to write to, usually ``"_metadata"`` in the root of
+            the output dataset
+        append: boolean
+            Whether or not to consolidate new metadata with existing (True)
+            or start from scratch (False)
+        **kwargs:
+            Other keywords as needed by the engine
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def write_partition(
+        df, path, fs, filename, partition_on, return_metadata, **kwargs
+    ):
+        """
+        Output a partition of a dask.DataFrame. This will correspond to
+        one output file, unless partition_on is set, in which case, it will
+        correspond to up to one file in each sub-directory.
+
+        Parameters
+        ----------
+        df
+        path
+        fs
+        filename
+        partition_on
+        kwargs
+        return_metadata : bool
+            Whether to return list of instances from this write, one for each
+            output file. These will be passed to write_metadata if an output
+            metadata file is requested.
```

Comment:
@martindurant are there cases where we wouldn't want this?  My gut reaction is not to give this option.  I may not be seeing the situations that you are though.

Reply:
We have heard of cases where gathering all the metadata results in a very big metadata file and large memory use on the worker that is to write it. In those cases where the metadata will not be useful (no column max/mins of much utility), I could imagine wanting to skip it. Of course, fastparquet always looks for the file... but there's a reason it became optional, and then no longer the default in spark.

Reply:
In that case I think that there is no harm in always returning metadata.  

If we don't want to write it then core can simply decide not to make the final task.  The return values from these tasks will simply be forgotten, without added cost.  There is one fewer keyword argument to think about and the return type is more predictable.  Thoughts?

Reply:
OK, that makes sense, but then in core
```
out = delayed(engine.write_metadata)(parts, meta, fs, path, **kwargs)
```
doesn't actually need to depend on the `parts` at all, when not writing the metadata file. I don't think there's any reason that the `_common_metadata` file (the one without the chunk specifics) need to be written last.

Reply:
Yes, that makes sense to me.  It's also a tiny bit nicer from a scheduling perspective if there isn't a final task.  

However, I'm guessing that your point is that this might be more complex for users, especially if they specify the `compute=False` keyword.  Now we're sending them a bunch of delayed objects, rather than just one.  We could resolve this by mapping a nullifying function across the outputs before sending to a final token function

```python
outs = [dask.delayed(lambda (fn, metadata): fn)(out) for out in outs]
out = dask.delayed(outs)
```

Regardless though, this bit of complexity seems like it should rest on Dask DataFrame's shoulders, and not necessarily on the Engines'.

Reply:
Might that solution not possibly cause the (potentially large) metadata objects to be copied around the cluster? Maybe you would expect the small lambda task to get fused.

Visually, it can be nice for users to have a "finalise" task to any big operation, as shown on the dashboard, but we don't insist on this elsewhere and I don't view it as necessary.

Reply:
> Might that solution not possibly cause the (potentially large) metadata objects to be copied around the cluster? Maybe you would expect the small lambda task to get fused.

It's very unlikely.  First, they might get fused in optimization.  Second, even if they didn't they're very fast, so the scheduler is unlikely to move them away from data.  Third, even if the occasional task did get moved, it's not a ton of data to move to a peer.  The challenge arose when they all had to move to one machine.  

> Visually, it can be nice for users to have a "finalise" task to any big operation, as shown on the dashboard, but we don't insist on this elsewhere and I don't view it as necessary.

 My guess is that if they say `to_parquet()` and `compute=True` then we just compute things without making a finalize task (this is easier on the system and simpler).  If they say `compute=False` then we add the finalize task (and any nullifying tasks) and give them a delayed object.  This is slightly less clean, but hopefully is fine, and is also uncommon.

Reply:
> and is also uncommon.

That's probably what makes it fine - if clearly described in the docs, of course.